app/gui2/src/util/ast/abstract.ts

```diff
@@ -0,0 +1,165 @@
+ * map are used to produce modifications to the Yjs data structures, and the overlay is cleared. By using the Yjs
+ * notifications even for local changes, we avoid needing separate tree-update mechanisms for local and remote changes.
+ *
+ * [In the initial implementation, until the new AST is used for synchronization, local edits can be committed by
+ * simply moving dirty nodes into the committed node map. Remote edits can recreate the committed-node map by parsing
+ * the new code.]
+ */
+
+import { Token,Tree } from '@/generated/ast'
+import { parseEnso } from '@/util/ast'
+import type { LazyObject } from '@/util/parserSupport'
+
+declare const brandAbstractNodeId: unique symbol
+export type AbstractNodeId = number & { [brandAbstractNodeId]: never }
+
+let nextNodeId = 0
+function newNodeId(): AbstractNodeId {
+  const id = nextNodeId
+  nextNodeId++
+  return id as AbstractNodeId
+}
+
+type NodeChild = { whitespace: string | undefined; node: AbstractNodeId | string }
+
+type AbstractNode = {
+  children: NodeChild[]
+}
+
+type NodeMap = Map<AbstractNodeId, AbstractNode>
+
+function abstract(
+  tree: Tree,
+  nodes: NodeMap,
+  source: string,
+): [AbstractNodeId, string | undefined] {
+  const children: NodeChild[] = []
+  const visitor = (child: LazyObject) => {
+    if (Tree.isInstance(child)) {
+      const [childId, whitespace] = abstract(child, nodes, source)
+      children.push({ whitespace, node: childId })
+    } else if (Token.isInstance(child)) {
+      const whitespace = source.substring(
+        child.whitespaceStartInCodeBuffer,
+        child.whitespaceStartInCodeBuffer + child.whitespaceLengthInCodeBuffer,
+      )
+      const node = source.substring(
+        child.startInCodeBuffer,
+        child.startInCodeBuffer + child.lengthInCodeBuffer,
+      )
+      children.push({ whitespace, node })
+    } else {
+      child.visitChildren(visitor)
+    }
+  }
+  tree.visitChildren(visitor)
+  const whitespace = source.substring(
+    tree.whitespaceStartInCodeParsed,
+    tree.whitespaceStartInCodeParsed + tree.whitespaceLengthInCodeParsed,
+  )
+  const id = newNodeId()
+  const node = { children }
+  nodes.set(id, node)
+  return [id, whitespace]
+}
+
+type Span = [number, number]
+type InfoMap = Map<Span, AbstractNodeId>
+
+function print(root: AbstractNodeId, nodes: NodeMap, info: InfoMap, offset?: number): string {
+  let node = nodes.get(root)
+  if (node == null) throw new Error('missing node?')
+  let result = ''
+  for (const child of node.children) {
+    if (child.whitespace != null) {
+      result += child.whitespace
+    } else if (result.length != 0) {
+      result += ' '
+    }
+    if (typeof child.node == 'string') {
+      result += child.node
+    } else {
+      result += print(child.node, nodes, info, result.length)
+    }
+  }
+  return result
+}
+
+if (import.meta.vitest) {
+  const { test, expect } = import.meta.vitest
+
+  const source = 'foo bar+baz'
```

Comment:
Let's move that test into its own test file. We will surely have more than that added later as well.